### PR TITLE
Set up the Home screen - Contacts lists

### PR DESCRIPTION
**Pull request by yostyle**
#5

### CODE REVIEW

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    private fun buildlocalContacts(currentState: TchapContactListViewState) {
+        val userList = currentState.filteredRoomSummaries.mapNotNull { item ->
+            item.directUserId?.let {
+                User(it, item.displayName, item.avatarUrl)
+            }
+        }
+        val tchapContactList: MutableList<Any> = userList.toMutableList()
+
+        tchapContactList.addAll(currentState.filteredLocalContacts.filter { mappedContact ->
+            mappedContact.emails.any { it.matrixId != null && !userList.any { user -> user.userId == it.matrixId } }
```
Comment (Florian14) [inline vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt:87]: I think you can simplify like this:
mappedContact.emails
    .filterNotNull { it.matrixId }
    .any { userList.none { user -> user.userId == it.matrixId }

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+
+                is MappedContact -> {
+                    contactItem {
+                        id(item.id)
+                        mappedContact(item)
+                        avatarRenderer(avatarRenderer)
+                    }
+                    item.emails
+                            .forEachIndexed { index, it ->
+                                if (it.matrixId == null) return@forEachIndexed
```
Comment (Florian14) [inline vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt:141]: This should work also:
`it.matrixId ?: return@forEachIndexed`

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+//                                callback?.onItemClick(item)
+//                            }
+//                        }
+//                    }
+//                }
+//    }
+
+    private fun buildDirectoryUsers(directoryUsers: List<User>, searchTerms: String, ignoreIds: List<String>) {
+        val toDisplay = directoryUsers
+                .filter { !ignoreIds.contains(it.userId) && it.userId != session.myUserId }
```
Comment (Florian14) [inline vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt:186]: What do you think about: `filterNot { ignoreIds.contains(it.userId) || it.userId == session.myUserId }` ?

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt (hunk 2 above)
Comment (yostyle) [reply vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt:141]: Yes, this code comes from element. The result is the same.

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt (hunk 3 above)
Comment (yostyle) [reply vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt:186]: Yes I prefer your proposal. Thanks

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt (hunk 1 above)
Comment (yostyle) [reply vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListController.kt:87]: Why not but in your case you iterate two times (filterNotNull + any) on the same elements.

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListAction.kt [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ */
+
+package fr.gouv.tchap.features.home.contact.list
+
+import im.vector.app.core.platform.VectorViewModelAction
+
+sealed class TchapContactListAction : VectorViewModelAction {
+    data class SearchUsers(val value: String) : TchapContactListAction()
+    object ClearSearchUsers : TchapContactListAction()
+    object ComputeMatrixToLinkForSharing : TchapContactListAction()
```
Comment (giomfo) [inline vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListAction.kt:24]: This action `ComputeMatrixToLinkForSharing` can be removed and the related code (`OpenShareMatrixToLink`) too

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewModel.kt [hunk 5]
```diff
@@ -0,0 +1,301 @@
...
+
+    private fun updateFilteredContacts() = withState { state ->
+        val filteredMappedContacts = mappedContacts
+                .filter { it.displayName.contains(state.searchTerm, true) }
+                .filter { contactModel -> contactModel.emails.any { it.matrixId != null } || contactModel.msisdns.any { it.matrixId != null }
+                }
+
+        val filteredRoomSummaries = state.roomSummaries.invoke()
+                ?.filter { it.displayName.contains(state.searchTerm, true) }
+                ?.filter { user -> user.directUserId != null }
```
Comment (giomfo) [inline vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewModel.kt:184]: user?
don't you think `summary` or `item` would be better?

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,301 @@
...
+                .orEmpty()
+
+        setState {
+            copy(
+                    filteredLocalContacts = filteredMappedContacts,
+                    filteredRoomSummaries = filteredRoomSummaries
```
Comment (giomfo) [inline vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewModel.kt:190]: I would like to check this with you

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewModel.kt [hunk 7]
```diff
@@ -0,0 +1,301 @@
...
+                    } else {
+                        val searchObservable = session.rx()
+                                .searchUsersDirectory(search, 50, state.excludedUserIds.orEmpty())
+                        // If it's a valid user id try to use Profile API
+                        // because directory only returns users that are in public rooms or share a room with you, where as
+                        // profile will work other federations
+                        if (!MatrixPatterns.isUserId(search)) {
+                            searchObservable
+                        } else {
+                            val profileObservable = session.rx().getProfileInfo(search)
```
Comment (giomfo) [inline vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewModel.kt:251]: I'm not sure we want this profileObservable here in Tchap

File: vector/src/main/java/im/vector/app/features/home/HomeActivityViewModel.kt [hunk 8]
```diff
@@ -207,6 +208,7 @@ class HomeActivityViewModel @AssistedInject constructor(
                                     false
                             )
                     )
+                    updateIdentityServer(session)
```
Comment (giomfo) [inline vector/src/main/java/im/vector/app/features/home/HomeActivityViewModel.kt:211]: I don't think this is the right way to set up the id server
We should check whether  the id server is defined or not (without depending on cross signing)

File: vector/src/main/java/im/vector/app/features/homeserver/HomeServerCapabilitiesViewModel.kt [hunk 9]
```diff
@@ -51,8 +52,11 @@ class HomeServerCapabilitiesViewModel @AssistedInject constructor(
     companion object : MvRxViewModelFactory<HomeServerCapabilitiesViewModel, HomeServerCapabilitiesViewState> {
         @JvmStatic
         override fun create(viewModelContext: ViewModelContext, state: HomeServerCapabilitiesViewState): HomeServerCapabilitiesViewModel? {
-            val fragment: UserListFragment = (viewModelContext as FragmentViewModelContext).fragment()
-            return fragment.homeServerCapabilitiesViewModelFactory.create(state)
+            val context = viewModelContext as FragmentViewModelContext
+            val factory = (context.fragment as? UserListFragment)?.homeServerCapabilitiesViewModelFactory
```
Comment (giomfo) [inline vector/src/main/java/im/vector/app/features/homeserver/HomeServerCapabilitiesViewModel.kt:56]: In which case the UserListFragment is expected?

**Review by giomfo — COMMENTED** [review]
FYI the compilation fails on your branch because `@color/vector_tchap_text_color_dark`is undefined

File: vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewModel.kt (hunk 7 above)
Comment (giomfo) [reply vector/src/main/java/fr/gouv/tchap/features/home/contact/list/TchapContactListViewModel.kt:251]: After thinking, we can remove the profileObs

**Review by giomfo — COMMENTED** [review]
The code looks good to me, but I observed some issues when I ran it with my account.
Some contacts are missing (they should come from my DMs)
Some others should be hidden (they are deactivated)

I know how to fix the second point, but I don't see the reason of the first one for the moment

**Review by giomfo — APPROVED** [review]
SGTM

**Pull request merged**